ddtrace/propagation/http.py

```diff
@@ -15,16 +23,279 @@
+            headers[HTTP_HEADER_B3_FLAGS] = "1"
+
+
+def _extract_b3(headers):
+    # type: (Dict[str, str]) -> Optional[Context]
+    trace_id_val = _extract_header_value(
+        POSSIBLE_HTTP_HEADER_B3_TRACE_IDS,
+        headers,
+    )
+    if trace_id_val is None:
+        return None
+
+    span_id_val = _extract_header_value(
+        POSSIBLE_HTTP_HEADER_B3_SPAN_IDS,
+        headers,
+    )
+    sampled = _extract_header_value(
+        POSSIBLE_HTTP_HEADER_B3_SAMPLEDS,
+        headers,
+    )
+    flags = _extract_header_value(
+        POSSIBLE_HTTP_HEADER_B3_FLAGS,
+        headers,
+    )
+
+    # Try to parse values into their expected types
+    try:
+        # DEV: We are allowed to have only x-b3-sampled/flags
+        # DEV: Do not allow `0` for trace id or span id, use None instead
+        trace_id = None
+        span_id = None
+        if trace_id_val is not None:
+            trace_id = _b3_id_to_dd_id(trace_id_val) or None
+        if span_id_val is not None:
+            span_id = _b3_id_to_dd_id(span_id_val) or None
+
+        sampling_priority = None
+        if sampled is not None:
+            if sampled == "0":
+                sampling_priority = AUTO_REJECT
+            elif sampled == "1":
+                sampling_priority = AUTO_KEEP
+        if flags == "1":
+            sampling_priority = USER_KEEP
+
+        return Context(
+            trace_id=trace_id,
+            span_id=span_id,
+            sampling_priority=sampling_priority,
+        )
+    except (TypeError, ValueError):
+        log.debug(
+            "received invalid x-b3-* headers, " "trace-id: %r, span-id: %r, sampled: %r, flags: %r",
+            trace_id_val,
+            span_id_val,
+            sampled,
+            flags,
+        )
+    return None
+
+
+def _inject_b3_single_header(span_context, headers):
+    # type: (Context, Dict[str, str]) -> None
+    if span_context.trace_id is None or span_context.span_id is None:
+        log.debug("tried to inject invalid context %r", span_context)
+        return
+
+    single_header = "{}-{}".format(_dd_id_to_b3_id(span_context.trace_id), _dd_id_to_b3_id(span_context.span_id))
+    sampling_priority = span_context.sampling_priority
+    if sampling_priority is not None:
+        if sampling_priority == 0:
+            single_header += "-0"
+        elif sampling_priority == 1:
+            single_header += "-1"
+        elif sampling_priority > 1:
+            single_header += "-d"
+    headers[HTTP_HEADER_B3_SINGLE] = single_header
+
+
+def _extract_b3_single_header(headers):
+    # type: (Dict[str, str]) -> Optional[Context]
+    single_header = _extract_header_value(POSSIBLE_HTTP_HEADER_B3_SINGLE_HEADER, headers)
+    if not single_header:
+        return None
+
+    trace_id = None
+    span_id = None
+    sampled = None
+
+    parts = single_header.split("-")
+    trace_id_val = None
+    span_id_val = None
+    if len(parts) == 1:
+        sampled = parts[0]
```

Comment:
🙁 
~~~ python
a = [None]
%timeit b, = a
%timeit b = a[0]
~~~

~~~
19.9 ns ± 0.195 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
30.3 ns ± 0.366 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
~~~